nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/StandardNiFiServiceFacade.java

```diff
@@ -2939,6 +2939,115 @@ public ControllerServiceEntity updateControllerService(final Revision revision,
+            // move to parent process group
+            newParent = oldParentGroup.getParent();
+            newParent.addControllerService(controllerService);
+
+            // unset any references the controller services has to other controller services that are now out of scope
+            Map<String, String> updatedProps = new HashMap<>();
+            Set<Map.Entry<PropertyDescriptor, PropertyConfiguration>> properties = controllerService.getProperties().entrySet();
+            for (var prop : properties) {
+                var value = prop.getValue();
+                if (value !=null) {
+                    ControllerServiceNode controller;
+                    try {
+                        controller = controllerServiceDAO.getControllerService((value.getRawValue()));
+                    } catch (Exception e){
+                        continue;
+                    }
+                    if (controller != null) {
+                        if (!hasProcessGroup(controller.getProcessGroup(), newParent.getIdentifier())) {
+                            controller.removeReference(controllerService, prop.getKey());
+                            updatedProps.put(prop.getKey().getName(), null);
+                        }
+                    }
+                }
+            }
+            if (!updatedProps.isEmpty())
+                controllerService.setProperties(updatedProps, true, Collections.emptySet());
+
+        } else {
+            // move to child process group
+            newParent = oldParentGroup.getProcessGroup(newProcessGroupID);
+            newParent.addControllerService(controllerService);
+
+            // unset any references for processors that are outside the new scope
+            for (ComponentNode node : referencedComponents) {
+                if (!hasProcessGroup(newParent, node.getProcessGroupIdentifier())) {
+                    Set<Map.Entry<PropertyDescriptor, PropertyConfiguration>> properties = node.getProperties().entrySet();
+                    Map<String, String> updatedProps = new HashMap<>();
+                    for (Map.Entry<PropertyDescriptor, PropertyConfiguration> prop : properties) {
+                        final PropertyConfiguration value = prop.getValue();
+                        if (value != null && value.getRawValue().equals(controllerService.getIdentifier())) {
+                            controllerService.removeReference(node, prop.getKey());
+                            node.getComponent().onPropertyModified(prop.getKey(), controllerService.getIdentifier(), null);
+                            updatedProps.put(prop.getKey().getName(), null);
+                        }
+                    }
+                    node.setProperties(updatedProps, true, Collections.emptySet());
+                }
+            }
+        }
+
+        return controllerService;
+    }
+    private boolean hasProcessGroup(ProcessGroup root, String id){
```

Comment:
This method is unnecessary. There already exists a method on `ProcessGroup` named `findProcessGroup`. So you can simply use `root.findProcessGroup(id) != null`

Reply:
Removed method.